supervisor/dbus/udisks2/__init__.py

```diff
@@ -0,0 +1,99 @@
+"""Interface to UDisks2 over D-Bus."""
+from dataclasses import dataclass
+import logging
+from typing import TypedDict
+
+from dbus_fast.aio import MessageBus
+from dbus_fast.signature import Variant
+from typing_extensions import NotRequired
+
+from ...exceptions import DBusError, DBusInterfaceError
+from ..const import (
+    DBUS_ATTR_SUPPORTED_FILESYSTEMS,
+    DBUS_IFACE_FILESYSTEM,
+    DBUS_IFACE_UDISKS2_MANAGER,
+    DBUS_NAME_UDISKS2,
+    DBUS_OBJECT_UDISKS2,
+)
+from ..interface import DBusInterfaceProxy, dbus_property
+from ..utils import dbus_connected
+from .block import UDisks2Block
+from .filesystem import UDisks2Filesystem
+
+_LOGGER: logging.Logger = logging.getLogger(__name__)
+
+UDisks2StandardOptionsDataType = TypedDict(
+    "UDisks2StandardOptionsDataType",
+    {"auth.no_user_interaction": NotRequired[bool]},
+)
+
+
+@dataclass
+class UDisks2StandardOptions:
+    """UDisks2 standard options."""
+
+    auth_no_user_interaction: bool | None
+
+    @staticmethod
+    def from_dict(data: UDisks2StandardOptionsDataType) -> "UDisks2StandardOptions":
+        """Create UDisks2StandardOptions from dict."""
+        return UDisks2StandardOptions(
+            auth_no_user_interaction=data.get("auth.no_user_interaction"),
+        )
+
+    def to_dict(self) -> dict[str, Variant]:
+        """Return dict representation."""
+        data = {"auth.no_user_interaction": Variant("b", self.auth_no_user_interaction)}
+        return {k: v for k, v in data.items() if v.value is not None}
+
+
+class UDisks2(DBusInterfaceProxy):
+    """Handle D-Bus interface for UDisks2.
+
+    http://storaged.org/doc/udisks2-api/latest/
+    """
+
+    name: str = DBUS_NAME_UDISKS2
+    bus_name: str = DBUS_NAME_UDISKS2
+    object_path: str = DBUS_OBJECT_UDISKS2
+    properties_interface: str = DBUS_IFACE_UDISKS2_MANAGER
+
+    async def connect(self, bus: MessageBus):
+        """Connect to D-Bus."""
+        try:
+            await super().connect(bus)
+        except DBusError:
+            _LOGGER.warning("Can't connect to udisks2")
+        except DBusInterfaceError:
+            _LOGGER.warning(
+                "No udisks2 support on the host. Host control has been disabled."
+            )
+
+    @property
+    @dbus_property
+    def supported_filesystems(self) -> list[str]:
+        """Return list of supported filesystems."""
+        return self.properties[DBUS_ATTR_SUPPORTED_FILESYSTEMS]
+
+    @dbus_connected
+    async def get_block_devices(self) -> list[UDisks2Block]:
+        """Return list of all block devices."""
+        devices: list[UDisks2Block] = []
+        for block_device in await self.dbus.Manager.call_get_block_devices({}):
+            device = UDisks2Block(block_device)
+            await device.connect(self.dbus.bus)
```

Comment:
Based on the current code, a signal listener is going to be set up for each of these block devices (and each of the filesystems below). But these appear to be transient, you pull info on each device in the list, return it and then the objects are tossed.

We should make one of the following changes:

1. Pull the list of block devices and filesystems and block devices during `update` and cache those lists. Return the cached list here and during future updates call `connect` on new devices/filesystems, `update` on existing ones and `disconnect` on ones that have disappeared. See here for an example of that approach:
https://github.com/home-assistant/supervisor/blob/30932a83f8dcbb6261823b468aeb6e1d28b4a1cd/supervisor/dbus/network/__init__.py#L177-L227
 
2. Add `sync_properties = false` to the class fields of `UDisks2Block` and `UDisks2Filesystem` so the signal listeners are not set up on `connect`. Then you can freely make and toss these devices every time this method is called. Like what we do for access points currently:
https://github.com/home-assistant/supervisor/blob/30932a83f8dcbb6261823b468aeb6e1d28b4a1cd/supervisor/dbus/network/wireless.py#L54-L65

But we need to do one of those two otherwise we risk leaking the listeners when the objects are garbage collected.